d2/src/main/java/com/linkedin/d2/xds/XdsClient.java

```diff
@@ -0,0 +1,311 @@
+package com.linkedin.d2.xds;
+
+import com.google.protobuf.Value;
+import io.envoyproxy.envoy.config.endpoint.v3.LocalityLbEndpoints;
+import io.envoyproxy.envoy.config.route.v3.VirtualHost;
+import io.envoyproxy.envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+
+public abstract class XdsClient
+{
+  private static final String ADS_TYPE_URL_LDS = "type.googleapis.com/envoy.config.listener.v3.Listener";
+  private static final String ADS_TYPE_URL_RDS = "type.googleapis.com/envoy.config.route.v3.RouteConfiguration";
+  private static final String ADS_TYPE_URL_CDS = "type.googleapis.com/envoy.config.cluster.v3.Cluster";
+  private static final String ADS_TYPE_URL_EDS = "type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment";
+
+  /**
+   * Watcher interface for a single requested xDS resource.
+   */
+  interface ResourceWatcher
+  {
+    // TODO: Handle onError and onResourceNotExist
+  }
+
+  interface LdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(LdsUpdate update);
+  }
+
+  interface RdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(RdsUpdate update);
+  }
+
+  interface CdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(CdsUpdate update);
+  }
+
+  interface EdsResourceWatcher extends ResourceWatcher
+  {
+    void onChanged(EdsUpdate update);
+  }
+
+  interface ResourceUpdate
+  {
+  }
+
+  static final class LdsUpdate implements ResourceUpdate
+  {
+    // Http level api listener configuration.
+    final HttpConnectionManager _httpConnectionManager;
```

Comment:
Better to add at least listener name, it will facilitate debugging in the future